### PR TITLE
Added framework for DeleteLocalVolume gRPC call + synchronized delete annotation

### PR DESCRIPTION
**Pull request by ChristianAtDell**
# Description
New annotation for synchronized deletion of backend volumes has been added, and logic to support this and a new gRPC call in the replicator sidecar (DeleteLocalVolume) have been added as well. 

Driver-level implementation of DeleteLocalVolume to follow. PR is set as draft for now.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| [665](https://github.com/dell/csm/issues/665) |

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

Usage of the new gRPC call and annotation has been tested with the PowerStore CSI driver. Annotations were applied and edited as expected, and remote PVs were deleted as expected.

### CODE REVIEW

File: controllers/common.go [hunk 1]
```diff
@@ -219,6 +219,10 @@ var (
 	ResourceRequest string
 	// DeletionRequested annotation which will be set to the PV on the object deletion
 	DeletionRequested string
+	// SynchronizedDeletionStatus indicates what state the PV's synchronized deletion process is in
+	// requested: local PV has been issued a delete command and needs to run DeleteRemoteVolume.
+	// complete: DeleteRemoteVolume has completed. Placed on local PV by replicator sidecar, propogated to remote PV via controller.
+	SynchronizedDeletionStatus string
```
[the remark this answers is not in the document]
Comment (ChristianAtDell) [reply controllers/common.go:225]: I am of the opinion that 'sync' carries with it the connotation of an array-to-array sync, while this is meant to *synchronize* an operation between two clusters. I personally think the more verbose 'synchronized' fits better here.

File: controllers/csi-replicator/persistentvolume_controller.go [hunk 2]
```diff
@@ -141,15 +141,29 @@ func (r *PersistentVolumeReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		}
 	}
 
-	log.V(common.DebugLevel).Info("Checking if PV protection complete annotation is applied")
-
-	if _, ok := pv.Annotations[controller.PVProtectionComplete]; !ok {
```
[the remark this answers is not in the document]
Comment (ChristianAtDell) [reply controllers/csi-replicator/persistentvolume_controller.go:146]: Good question... I don't recall deleting this code block. Adding it back.

File: controllers/constants.go [hunk 3]
```diff
@@ -87,6 +87,8 @@ const (
 	resourceRequest = "/resourceRequest"
 	// DeletionRequested annotation to set for the deleted object
 	deletionRequested = "/deletionRequested"
+	// SynchronizedDeletionStatus annotation to communicate RemoteVolumeDeletion between cluster
```
Comment (santhoshatdell) [inline controllers/constants.go:90]: Comment needs update.

File: controllers/csi-replicator/persistentvolume_controller.go [hunk 4]
```diff
@@ -142,14 +142,35 @@ func (r *PersistentVolumeReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
+		}
+	}
+
+	log.V(common.DebugLevel).Info("Checking if synchronized deletion annotation is applied")
+	syncDeleteStatus, ok := pv.Annotations[controller.SynchronizedDeletionStatus]
+	if ok && syncDeleteStatus == "requested" {
+		log.V(common.InfoLevel).Info("Synchronized Deletion requested by annotation, deleting local backend volume")
+		volumeHandle := pv.Spec.PersistentVolumeSource.CSI.VolumeHandle
+		if volumeHandle == "" {
+			log.Error(err, "Failed to retrieve the PV's remote volume handle for deletion")
```
Comment (santhoshatdell) [inline controllers/csi-replicator/persistentvolume_controller.go:158]: remote -> local/none

File: test/mock-server/server/server.go [hunk 5]
```diff
@@ -130,6 +130,13 @@ func (s *Replication) CreateRemoteVolume(ctx context.Context, in *replication.Cr
 	return out, err
 }
 
+// DeleteLocalVolume calls stub for DeleteLocalVolume
+func (s *Replication) DeleteLocalVolume(ctx context.Context, in *replication.DeleteLocalVolumeRequest) (*replication.DeleteLocalVolumeResponse, error) {
+	out := &replication.DeleteLocalVolumeResponse{}
+	err := FindStub("Replication", "DeleteLocalVolume", in, out)
+	return out, err
+}
+
```
Comment (santhoshatdell) [inline test/mock-server/server/server.go:139]: Update copyrights year. Also check other files.

**Review by santhoshatdell — APPROVED** [review]

**Review by santhoshatdell — DISMISSED** [review]

**Review by santhoshatdell — APPROVED** [review]

**Review by delldubey — APPROVED** [review]
lgtm

**Review by santhoshatdell — APPROVED** [review]

**Pull request closed without merging**